test/test_autograd_onboarding_lab.py

```diff
@@ -0,0 +1,70 @@
+import torch
+from torch.autograd import Function, gradcheck, gradgradcheck
+from torch.testing._internal.common_utils import TestCase, run_tests
+
+class TestOnboardinngAttn(TestCase):
+    class Onboarding_Attn(Function):
+
+        @staticmethod
+        def forward(ctx, q, k, v):
+            ''' Inputs
+                q: m * n
+                k: m * n
+                v: m * p
+                Outputs
+                a: m * m
+                o: m * p
+            '''
+            # check that inputs are 2D and check sizes of input
+            assert q.dim() == 2 and k.dim() == 2 and v.dim() == 2, "all inputs must have dim=2"
+            assert q.size()[1] == k.size()[1], "1st dim of q and 1st dim of k must have the same size"
+            assert k.size()[0] == v.size()[0], "0th dim of k and 0th dim of v must have the same size"
+
+            x = torch.matmul(q, k.transpose(0, 1))
+            a = torch.tanh(x)
+            o = torch.matmul(a, v)
+            ctx.save_for_backward(q, k, v, a)
```

Comment:
Note that you don't have to save intermediary results in this way, only input/output need to be here.
But for simplicity it's ok to treat all of them the same way here.

Reply:
Just to make sure I'm understanding, I should save q, k, v, a, o here because they are the inputs + outputs and x should not be saved here because gradients will not flow in backward, is that right?

Reply:
double backward will indeed be problematic for x because gradient won't flow properly.